aw-transform/src/flood.rs

```diff
@@ -40,23 +67,78 @@ pub fn flood(events: Vec<Event>, pulsetime: chrono::Duration) -> Vec<Event> {
+            //         logger.warning(
+            //             "Gap was of negative duration but could be safely merged ({}s). This message will only show once per batch.".format(
+            //                 gap.total_seconds()
+            //             )
+            //         )
+            //         warned_about_negative_gap_safe = True
+
+            // If data is same, we can safely merge them.
             if e1.data == e2.data {
-                if chrono::Duration::seconds(0) > gap && !warned_negative_gap_safe {
+                // Gap was negative and could be safely merged
+                if !warned_negative_gap_safe {
                     warn!("Gap was of negative duration ({}s), but could be safely merged. This error will only show once per batch.", gap);
                     warned_negative_gap_safe = true;
                 }
+                let start = std::cmp::min(e1.timestamp, e2.timestamp);
+                let end = std::cmp::max(e1.calculate_endtime(), e2.calculate_endtime()); // e2 isn't guarranteed to end last
+                e1.timestamp = start;
+                e1.duration = end - start;
+                // Drop next event since it is merged/flooded into e1
+                e1_iter.next();
+                // Retry this event again to give it a change to merge e1
+                // with 'e3'
+                retry_e = Some(e1);
+                continue;
+            // If data differs and gap is negative beyond the trim thres, throw a warning
+            } else if gap < -negative_gap_trim_thres && !warned_negative_gap_unsafe {
+                // Events with negative gap but differing data cannot be merged safely
+                warn!("Gap was of negative duration and could NOT be safely merged ({}s). This warning will only show once per batch.", gap);
+                warned_negative_gap_unsafe = true;
+            }
+
+        // By now, we've ensured gap is non-negative, now we want to know if the gap is smaller
+        // than pulsetime, in which case we should fill it.
+        } else if gap < pulsetime {
+            // Python implementation:
+            //
+            // elif gap < -negative_gap_trim_thres and not warned_about_negative_gap_unsafe:
+            //     # Events with negative gap but differing data cannot be merged safely
+            //     logger.warning(
+            //         "Gap was of negative duration and could NOT be safely merged ({}s). This warning will only show once per batch.".format(
+            //             gap.total_seconds()
+            //         )
+            //     )
+            //     warned_about_negative_gap_unsafe = True
+
+            // Ensure that gap is actually non-negative here, at least in tests
+            debug_assert!(gap >= chrono::Duration::seconds(0));
+
+            // If data is the same, we should merge them.
+            if e1.data == e2.data {
                 // Choose the longest event and set the endtime to it
```

Comment:
```suggestion
                // Set the endtime of e1 to the last endtime of [e1, e2]
```

Reply:
>very unlikely scenario

Unfortunately not too unlikely. Not sure if they appear due to some faulty queuing, heartbeat logic, or race condition, but they pop up every now and then...